crates/common/tedge_utils/src/atomic.rs

```diff
@@ -0,0 +1,125 @@
+//! Utilities for atomic writes.
+//!
+//! For deployment of configuration files, we need to create a file atomically because certain
+//! programs might watch configuration file for changes, so if it's not written atomically, then
+//! file might be only partially written and a program trying to read it may crash.
+//!
+//! Atomic write of a file consists of creating a temporary file in the same directory, filling it
+//! with correct content and permissions, and only then renaming the temporary into the destination
+//! filename. Because we're never actually writing into the file, we don't need to write permissions
+//! for the destination file, even if it exists. Instead we need only write/execute permissions to
+//! the directory file is located in unless the directory has a sticky bit set. Overwriting a file
+//! will also change its uid/gid/mode, if writing process euid/egid is different from file's
+//! uid/gid. To keep uid/gid the same, after the write we need to do `chown`, and to do it we need
+//! sudo.
+
+use std::io::ErrorKind;
+use std::io::Read;
+use std::os::unix::fs::fchown;
+use std::os::unix::fs::MetadataExt;
+use std::os::unix::fs::PermissionsExt;
+use std::path::Path;
+
+use anyhow::Context;
+
+/// Writes a file atomically and optionally sets its permissions.
+///
+/// Setting ownership of a file is a privileged operation so it needs to be run as root. If any of
+/// the filesystem operations fail due to not having permissions, the function will return an error.
+///
+/// If the file already exists, its content will be overwritten but its permissions will remain
+/// unchanged.
+pub fn write_file_atomic_set_permissions_if_doesnt_exist(
```

Comment:
```suggestion
pub fn write_file_atomic_with_permissions_if_new(
```

Reply:
I don't think a function name has to tell the whole story about that function ;-)